### PR TITLE
Refactor comm modes

### LINKED ISSUE #83: More idiomatic C++ for CommMode
From https://kokkosteam.slack.com/archives/C06AF7352LE/p1718145132652509?thread_ts=1718128444.322109&cid=C06AF7352LE

> why `send<CommMode>( args...)` instead of `send(CommMode(), args...)`
> I feel the tag type approach as argument is more common nowadays for stuff like ISO C++ library design?

### PR DESCRIPTION
**Pull request by dssgabriel**
Closes #83.

`KokkosComm::send`/`isend` can now be called like so, following the tag-type-as-argument approach, encouraged by ISO C++ library design guidelines:
```cpp
using namespace KokkosComm;

// Send with explicit "synchronous" communication mode and execution space:
send(ReadyCommMode(), Kokkos::DefaultExecutionSpace(), view, dest, tag, comm);

// Send with implicit "default" communication mode and explicit execution space:
send(Kokkos::DefaultExecutionSpace(), view, dest, tag, comm);

// Send with explicit communication mode and implicit "default" execution space:
send(SynchronousCommMode(), view, dest, tag, comm);

// Send with implicit "default" communication mode and "default" execution space:
send(view, dest, tag, comm);
```

`send` and `isend` functions use a newly introduced `CommunicationMode` concept that enforces correct typing of the specified communication mode.

---

Reviews are welcome! I'm not a C++ expert, and I am not sure this is the best way to do it.

Two questions as well:
1. is the namespace around the comm mode structures really necessary?
2. would it be better to have function specializations instead of dispatching in the lambda (like in the previous implementation)? It may be more verbose, but as I understand, it is the preferred way to do it.

While we figure this out, I am converting the PR to a draft.

---

Edit: added code example.

### CODE REVIEW

File: src/impl/KokkosComm_isend.hpp [hunk 1]
```diff
@@ -45,24 +46,25 @@ void isend(const SendView &sv, int dest, int tag, MPI_Comm comm, MPI_Request &re
...
 
   using KCT  = KokkosComm::Traits<SendView>;
   using KCPT = KokkosComm::PackTraits<SendView>;
 
-  auto mpi_isend_fn = [](void *mpi_view, int mpi_count, MPI_Datatype mpi_datatype, int mpi_dest, int mpi_tag,
-                         MPI_Comm mpi_comm, MPI_Request *mpi_req) {
-    if constexpr (SendMode == CommMode::Standard) {
+  auto mpi_isend_fn = [mode](void *mpi_view, int mpi_count, MPI_Datatype mpi_datatype, int mpi_dest, int mpi_tag,
+                             MPI_Comm mpi_comm, MPI_Request *mpi_req) {
+    if constexpr (std::is_same_v<decltype(mode), CommMode::Standard>) {
```
Comment (devreal) [inline src/impl/KokkosComm_isend.hpp]: Not sure this works. `decltype(mode)` is `const SendMode&` which is not the same as any of the comm modes. Replace `decltype(mode)` with `SendMode` and you should be good.

**Comment by dssgabriel** [comment]
C++ experts' reviews welcome! 🙂

**Review by devreal — DISMISSED** [review]

**Comment by cedricchevalier19** [comment]
@dssgabriel Can you add an example of the change in the description of the PR?

**Review by cedricchevalier19 — COMMENTED** [review]
I think you should update the documentation as part of this PR.

File: docs/api/core.rst [hunk 2]
```diff
@@ -116,28 +118,30 @@ Collective
...
+^^^^^^^^^^^^^^^^^^^
 
-.. cpp:enum-class:: KokkosComm::CommMode
+Structures to specify the mode of an operation. Buffered mode is not supported.
 
-    A scoped enum to specify the mode of an operation. Buffered mode is not supported.
+.. cpp:struct:: KokkosComm::StandardCommMode
 
-    .. cpp:enumerator:: KokkosComm::CommMode::Standard
+  Let the MPI implementation decides whether outgoing messages will be buffered. Send operations can be started whether or not a matching receive has been started. They may complete before a matching receive is started. Standard mode is non-local: successful completion of the send operation may depend on the occurrence of a matching receive.
```
Comment (cedricchevalier19) [inline docs/api/core.rst]:
```suggestion
  Let the MPI implementation decide whether outgoing messages will be buffered. Send operations can be started whether or not a matching receive has been started. They may complete before a matching begins. Standard mode is non-local: successful completion of the send operation may depend on the occurrence of a matching receive.
```

File: docs/api/core.rst [hunk 3]
```diff
@@ -116,28 +118,30 @@ Collective
...
+.. cpp:struct:: KokkosComm::ReadyCommMode
 
-    .. cpp:enumerator:: KokkosComm::CommMode::Synchronous
+  Send operations may be started only if the matching receive is already started.
 
-      Synchronous mode: Send operations complete successfully only if a matching receive is started, and the receive operation has started to receive the message sent.
+.. cpp:struct:: KokkosComm::DefaultCommMode
 
-    .. cpp:enumerator:: KokkosComm::CommMode::Default
+  Default mode aliases ``Standard`` mode, but lets users override the behavior of operations at compile-time using the ``KOKKOSCOMM_FORCE_SYNCHRONOUS_MODE`` pre-processor definition. This forces ``Synchronous`` mode for all "default-mode" operations, which can be useful for debugging purposes, e.g., asserting that the communication scheme is correct.
```
Comment (cedricchevalier19) [inline docs/api/core.rst]:
```suggestion
  The default mode is aliased as ``Standard``but lets users override the behavior of operations at compile-time using the ``KOKKOMM_FORCE_SYNCHRONOUS_MODE `` pre-processor definition. This forces ``Synchronous`` mode for all "default-mode" operations, which can be helpful for debugging purposes, e.g., asserting that the communication scheme is correct.
```

File: perf_tests/test_2dhalo.cpp [hunk 4]
```diff
@@ -22,9 +22,9 @@
 
 void noop(benchmark::State, MPI_Comm) {}
 
-template <typename Space, typename View>
-void send_recv(benchmark::State &, MPI_Comm comm, const Space &space, int nx, int ny, int rx, int ry, int rs,
-               const View &v) {
+template <typename Mode, typename Space, typename View>
```
Comment (cedricchevalier19) [inline perf_tests/test_2dhalo.cpp]:
```suggestion
template <CommunicationMode Mode, typename Space, typename View>
```

File: perf_tests/test_osu_latency_isendirecv.cpp [hunk 5]
```diff
@@ -21,11 +21,11 @@
 #include "test_utils.hpp"
 #include "KokkosComm.hpp"
 
-template <typename Space, typename View>
-void osu_latency_Kokkos_Comm_isendirecv(benchmark::State &, MPI_Comm comm, const Space &space, int rank,
-                                        const View &v) {
+template <typename Mode, typename Space, typename View>
```
Comment (cedricchevalier19) [inline perf_tests/test_osu_latency_isendirecv.cpp]:
```suggestion
template <CommunicationMode Mode, typename Space, typename View>
```

File: perf_tests/test_osu_latency_sendrecv.cpp [hunk 6]
```diff
@@ -21,10 +21,11 @@
 #include "test_utils.hpp"
 #include "KokkosComm.hpp"
 
-template <typename Space, typename View>
-void osu_latency_Kokkos_Comm_sendrecv(benchmark::State &, MPI_Comm comm, const Space &space, int rank, const View &v) {
+template <typename Mode, typename Space, typename View>
```
Comment (cedricchevalier19) [inline perf_tests/test_osu_latency_sendrecv.cpp]:
```suggestion
template <CommunicationMode Mode, typename Space, typename View>
```

File: perf_tests/test_sendrecv.cpp [hunk 7]
```diff
@@ -18,14 +18,14 @@
 
 #include "KokkosComm.hpp"
 
-template <typename Space, typename View>
-void send_recv(benchmark::State &, MPI_Comm comm, const Space &space, int rank, const View &v) {
+template <typename Mode, typename Space, typename View>
```
Comment (cedricchevalier19) [inline perf_tests/test_sendrecv.cpp]:
```suggestion
template <CommunicationMode Mode, typename Space, typename View>
```

File: unit_tests/test_isendrecv.cpp [hunk 8]
```diff
@@ -30,9 +31,9 @@ using ScalarTypes =
     ::testing::Types<float, double, Kokkos::complex<float>, Kokkos::complex<double>, int, unsigned, int64_t, size_t>;
 TYPED_TEST_SUITE(IsendRecv, ScalarTypes);
 
-template <KokkosComm::CommMode IsendMode, typename Scalar>
+template <typename IsendMode, typename Scalar>
```
Comment (cedricchevalier19) [inline unit_tests/test_isendrecv.cpp]:
```suggestion
template <CommunicationMode IsendMode, typename Scalar>
```

File: unit_tests/test_isendrecv.cpp [hunk 9]
```diff
@@ -61,9 +61,9 @@ void isend_comm_mode_1d_contig() {
   }
 }
 
-template <KokkosComm::CommMode IsendMode, typename Scalar>
+template <typename IsendMode, typename Scalar>
```
Comment (cedricchevalier19) [inline unit_tests/test_isendrecv.cpp]:
```suggestion
template <CommunicationMode IsendMode, typename Scalar>
```

File: unit_tests/test_sendrecv.cpp [hunk 10]
```diff
@@ -29,9 +30,9 @@ class SendRecv : public testing::Test {
 using ScalarTypes = ::testing::Types<int, int64_t, float, double, Kokkos::complex<float>, Kokkos::complex<double>>;
 TYPED_TEST_SUITE(SendRecv, ScalarTypes);
 
-template <KokkosComm::CommMode SendMode, typename Scalar>
+template <typename SendMode, typename Scalar>
```
Comment (cedricchevalier19) [inline unit_tests/test_sendrecv.cpp]:
```suggestion
template <CommunicationMode SendMode, typename Scalar>
```

File: unit_tests/test_sendrecv.cpp [hunk 11]
```diff
@@ -59,9 +60,9 @@ void send_comm_mode_1d_contig() {
   }
 }
 
-template <KokkosComm::CommMode SendMode, typename Scalar>
+template <typename SendMode, typename Scalar>
```
Comment (cedricchevalier19) [inline unit_tests/test_sendrecv.cpp]:
```suggestion
template <CommunicationMode SendMode, typename Scalar>
```

**Review by cedricchevalier19 — COMMENTED** [review]
The `CommunicationMode` concept is a good idea, use it everywhere!

File: docs/api/core.rst (hunk 2 above)
Comment (dssgabriel) [reply docs/api/core.rst]: > They may complete before a matching begins.

Accidentally removed "receive"?

File: docs/api/core.rst (hunk 3 above)
Comment (dssgabriel) [reply docs/api/core.rst]: > `KOKKOMM_FORCE_SYNCHRONOUS_MODE`

Voluntary abbreviation of `KOKKOSCOMM` into `KOKKOMM`?

Comment (cedricchevalier19) [reply docs/api/core.rst]: Oups, do not change the define!

Comment (dssgabriel) [reply docs/api/core.rst]: All good! I kept it as it was and only updated your other suggestions :+1:

**Review by cedricchevalier19 — APPROVED** [review]

File: src/impl/KokkosComm_isend.hpp [hunk 12]
```diff
@@ -93,4 +88,9 @@ KokkosComm::Req isend(const ExecSpace &space, const SendView &sv, int dest, int
   return req;
 }
 
+template <KokkosExecutionSpace ExecSpace, KokkosView SendView>
+Req isend(const ExecSpace &space, const SendView &sv, int dest, int tag, MPI_Comm comm) {
+  return isend(KokkosComm::DefaultCommMode(), space, sv, dest, tag, comm);
+}
+
```
Comment (cwpearson) [inline src/impl/KokkosComm_isend.hpp:95]: 👍

**Review by cwpearson — APPROVED** [review]
I like this better! Thanks for updating the docs.

**Pull request merged**